pytransform3d/rotations/_conversions.py

```diff
@@ -1383,6 +1383,122 @@ def extrinsic_euler_zxy_from_active_matrix(R, strict_check=True):
         R, unity, unitx, unitz, False, strict_check)[::-1]
 
 
+def euler_from_quaternion(q, i, j, k, extrinsic):
+    """General method to extract any Euler angles from quaternions.
+
+    Parameters
+    ----------
+    q : array-like, shape (4,)
+        Unit quaternion to represent rotation: (w, x, y, z)
+
+    i : int
+        Axis of the first rotation angle. A number between 0 and 2.
+
+    j : int
+        Axis of the second rotation angle. A number between 0 and 2.
+
+    k : int
+        Axis of the third rotation angle. A number between 0 and 2.
+
+    extrinsic : bool
+        Do we use extrinsic transformations? Intrinsic otherwise.
+
+    Returns
+    -------
+    euler_angles : array, shape (3,)
+        Extracted rotation angles in radians about the axes i, j, k in this
+        order. The first and last angle are normalized to [-pi, pi]. The middle
+        angle is normalized to either [0, pi] (proper Euler angles) or
+        [-pi/2, pi/2] (Cardan / Tait-Bryan angles).
+
+    References
+    ----------
+    Bernardes, Evandro; Viollet, Stephane: Quaternion to Euler angles
+    conversion: A direct, general and computationally efficient method,
+    https://doi.org/10.1371/journal.pone.0276302
+    """
+    q = check_quaternion(q)
+
+    i += 1
+    j += 1
+    k += 1
+
+    # The original algorithm assumes extrinsic convention. Hence, we swap
+    # the order of axes for intrinsic rotation.
+    if not extrinsic:
+        i, k = k, i
+
+    # Proper Euler angles rotate about the same axis in the first and last
+    # rotation. If this is not the case, they are called Cardan or Tait-Bryan
+    # angles and have to be handled differently.
+    proper_euler = i == k
+    if proper_euler:
+        k = 6 - i - j
+
+    sign = (i - j) * (j - k) * (k - i) // 2
+    a = q[0]
+    b = q[i]
+    c = q[j]
+    d = q[k] * sign
+
+    if not proper_euler:
+        a, b, c, d = a - c, b + d, c + a, d - b
+
+    # Equation 35
+    angle_j = math.acos(
```

Comment:
Out of curiosity, is there any reason to use the `math` module since you're already using `numpy`?

Reply:
It's faster for scalars. When operating on arrays, numpy is faster though.

```python
In [1]: import numpy as np

In [2]: import math

In [3]: %timeit np.arccos(np.random.rand())
1.34 µs ± 12.6 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [4]: %timeit math.acos(np.random.rand())
384 ns ± 5.94 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [5]: %timeit np.arctan2(np.random.rand(), np.random.rand())
1.88 µs ± 21.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [6]: %timeit math.atan2(np.random.rand(), np.random.rand())
689 ns ± 9.63 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```